src/lib.rs

```diff
@@ -0,0 +1,272 @@
+// or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING
+// BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+// DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+//! An immutable cactus stuck (also called a spaghetti stack or parent pointer tree). A cactus
+//! stack is a (possibly empty) node with a (possibly null) pointer to a parent node. Pushing
+//! and popping return new nodes rather than mutating existing nodes.
+//!
+//! ```
+//! use cactus::Cactus;
+//! let c = Cactus::new();
+//! assert!(c.is_empty());
+//! let c2 = c.push(1);
+//! assert_eq!(c2.len(), 1);
+//! assert_eq!(*c2.val().unwrap(), 1);
+//! let c3 = c2.pop().unwrap();
+//! assert!(c3.is_empty());
+//! ```
+//!
+//! From a given node one can create multiple sub-stacks:
+//!
+//! ```
+//! use cactus::Cactus;
+//! let c = Cactus::new().push(1);
+//! let c2 = c.push(2);
+//! let c3 = c.push(3);
+//! assert!(c2 != c3);
+//! assert_eq!(c2.iter().cloned().collect::<Vec<_>>(), [2, 1]);
+//! assert_eq!(c3.iter().cloned().collect::<Vec<_>>(), [3, 1]);
+//! ```
+
+use std::fmt;
+use std::rc::Rc;
+
+/// An immutable cactus stack node. May be empty or contain a value; may have a pointer to a parent
+/// or not.
+#[derive(Clone, Default)]
+pub struct Cactus<T> {
+    node: Option<Rc<Node<T>>>
+}
+
+#[derive(Clone)]
+struct Node<T> {
+    val: T,
+    parent: Option<Rc<Node<T>>>
+}
+
+impl<T> Cactus<T> {
+    /// Return an empty cactus stack node.
+    pub fn new() -> Cactus<T> {
+        Cactus{node: None}
+    }
+
+    /// Is this cactus stack node empty?
+    ///
+    /// # Examples
+    /// ```
+    /// use cactus::Cactus;
+    /// let c = Cactus::new();
+    /// assert!(c.is_empty());
+    /// let c2 = c.push(1);
+    /// assert!(!c2.is_empty());
+    /// ```
+    pub fn is_empty(&self) -> bool {
+        self.node.is_none()
+    }
+
+    /// How many items are there in this cactus stack?
+    pub fn len(&self) -> usize {
+        if self.is_empty() {
+            return 0;
+        }
+        let mut cur = self.pop().unwrap();
+        let mut l = 1;
+        while !cur.is_empty() {
+            l += 1;
+            cur = cur.pop().unwrap();
+        }
+        l
+    }
+
+    /// Create a new cactus stack node containing value `val` and parent `self`.
+    ///
+    /// # Examples
+    /// ```
+    /// use cactus::Cactus;
+    /// let c = Cactus::new();
+    /// let c2 = c.push(1);
+    /// let c3 = c2.push(2);
+    /// assert_eq!(c3.iter().cloned().collect::<Vec<_>>(), [2, 1]);
+    /// ```
+    pub fn push(&self, val: T) -> Cactus<T> {
```

Comment:
This can be simplified as well

```rust
    pub fn push(&self, val: T) -> Cactus<T> {
        Cactus {
            node: Some(Rc::new(Node {
                val,
                parent: self.node.clone(),
            })),
        }
    }
```